torch/csrc/jit/script/compiler.cpp

```diff
@@ -0,0 +1,444 @@
+      } else if (type == "b") {
+          return {createConstant(at::CPU(at::kByte).scalarTensor(val))};
+      } else if (type == "i") {
+          return {createConstant(at::CPU(at::kInt).scalarTensor(val))};
+      } else {
+        throw std::runtime_error("unknown const type " + type);
+      }
+  }
+
+  std::vector<Value*> emitNode(NodeKind kind,
+                               const std::vector<Value*> inputs,
+                               const std::unordered_map<std::string, TreeRef>& attributes,
+                               const size_t output_size) {
+      Node * n = def.graph->appendNode(def.graph->create(kind, output_size));
+      for (auto* input_value : inputs) {
+          n->addInput(input_value);
+      }
+      for (const auto& iter : attributes) {
+          const auto& name = Symbol(iter.first);
+          const auto& value = iter.second;
+          // TODO: handle non-float attributes
+          switch (value->kind()) {
+          case TK_CONST: {
+              auto v = value->tree(0)->doubleValue();
+              auto type = value->tree(1)->stringValue();
+              if (type == "f") {
+                  n->f_(name, v);
+              } else {
+                  n->i_(name, v);
+              }
+          } break;
+          case TK_LIST:
+              if (value->trees().size()) {
+                  std::vector<double> values{};
+                  for (const auto& tree : value->trees()) {
+                      values.push_back(tree->tree(0)->doubleValue());
+                  }
+                  if (value->trees()[0]->tree(1)->stringValue() == "f") {
+                      n->fs_(name, std::move(values));
+                  } else {
+                      n->is_(name, std::vector<int64_t>(values.begin(), values.end()));
+                  }
+              }
+              break;
+          }
+      }
+      return n->outputs();
+  }
+
+  // Desugars slice syntactic sugar tensor[begin:end] -> tensor.slice(begin, end).
+  std::vector<Value*> emitSlice(const SourceRange& range,
+                                TreeList&& inputs,
+                                const size_t output_size) {
+    const auto applyInputs =
+        Compound::create(TK_LIST, range, std::move(inputs));
+    const auto applyAttributes = Compound::create(TK_LIST, range, {});
+
+    return emitNode(Symbol("slice"), getValues(applyInputs->trees()), {}, output_size);
+  }
+
+  // Desugars gather syntactic sugar tensor[indices] -> tensor.gather(indices).
+  std::vector<Value*> emitGather(const SourceRange& range,
+                                TreeList&& inputs,
+                                const size_t output_size) {
+    const auto applyInputs =
+        Compound::create(TK_LIST, range, std::move(inputs));
+    const auto applyAttributes = Compound::create(TK_LIST, range, {});
+    return emitNode(Symbol("gather"), getValues(applyInputs->trees()), {}, output_size);
```

Comment:
This doesn't look right. `gather` isn't the right ATen op for e.g. `tensor[2]` (this should be `select`)